### ISSUE TITLE
Clarify Code base

### ISSUE THREAD

**Issue by smiths** [smiths/vdisp#122]
@EmilSoleymani in looking through the code, I find it isn't as clear as I had thought it would be.  I should have reviewed the code sooner, but we were focusing on getting the functionality working.  I think we can work during the last few days of your summer to clarify and document the code.  I can work on the SRS myself, but reverse engineering the design from the code will be a bit more challenging.  

I'll create a list of things we can look at to clarify the design and its documentation.  We can discuss the list of ideas through the issue tracker and during tomorrow's (Wednesday's) meeting.  If it makes sense to do so, we can then break this issue up into multiple simpler issues.

The InputParser module and the vdisp modules seem too big to me.  Parnas's definition of a module is that it is a work assignment for a single programmer.  The idea is that the programmer would be given the specification and then produce the module.  Ideally a module should provide one set of services and encapsulate one secret.  Some specific ideas for breaking these modules into smaller modules follows:

- Move the @enums in the InputParser module into their own module
- Move the Custom Exceptions in the InputParser module into their own module

As mentioned in #42, I think it would be worthwhile to consider splitting the input parser into two modules.  One module holds the data (the contents of the InputData struct), with access programs for modifying the state of this module.  The InputParser module is then only responsible for parsing the input file and using this information to populate the InputData module's state (data).  Although you do everything through the file, it is conceivable that the GUI could modify the state of the InputData module directly.  It doesn't really have to go through the intermediary of a file.  We also might in the future have an algorithm that modifies the InputData without using a file, possibly as part of an optimization algorithm.  Instead of modifying the state of InputData directly, the InputParser module would call the setters for the InputData module.

The InputData functions have far too many arguments.   It seems to be a setter for the entire state of the InputData.  This means that the interface is not minimal.  A minimal interface avoids access routines with two potentially independent services.  You can see more about quality criteria for interfaces on [Slide 17 of my Module Interface Design Lecture](https://gitlab.cas.mcmaster.ca/smiths/se2aa4_cs2me3/-/blob/master/Lectures/L16_ModuleIntefaceDesign/ModuleInterfaceDesign.pdf).  Isn't there a case where we would independently change the state of the input data?  What if a user just modifies one field in the GUI?  It feels like InputData should be split into more setters.  Maybe not one per state variable, but separate setters for each logically linked types of inputs.  This might also avoid the problem of essentially repeating the InputData over and over again.

The InputParser module seems to work from the GUI and from a file.  This violates the one module one secret rule.  The InputParser should either know the file format, or the GUI format, not both.

For testing each module should have its own set of test cases in a corresponding module.  The InputData module should have a set of tests for the setters.  The InputParser module should have its own set of tests.  Each of your behaviour modules should have its own set of tests, in their own file.  It appears that all of the tests are in the vdisp module?  Can we separate them into their own files.  At the very least, can we separate the test cases from the vdisp module?

The vdisp file could be further subdivided into smaller modules.  Some ideas include:

- putting the constants in their own file
- maybe the graphing could be moved to another module?

The code 

```
if PRINT_DEBUG
        println("Got an update: ", val)
    end
```
Is repeated many times in vdisp.  Could this be moved into a function?

The folder of modules named OutputFormat is confusing since the modules are related to behaviour.  Looking at the modules it seems like they have services for calculations and for output.  It would be better to have modules that calculate the outputs separate from the module that displays the outputs.  Do we have a module that stores the output state?  The calculation modules can populate the output state.  You can then take as many different views of this output state as you want, graphical views or file views.

We need to spend some time on documenting the design.  The UML diagram is a good start, but it doesn't give the details of the Model, Foundation and Tuple classes.  Ideally I would like to have the syntax section for each module summarized.  I don't think we have time for this though.  Maybe we can have one or two sentences for each module that summarizes its services and secret?  I'd also like a blurb that describes the design.  This would probably be clearest by describing the data flow.  How does the data flow from input to calculations to input?  Where is the state information stored?  That is, where are the inputs and outputs stored?  What modules are responsible for transforming and viewing the inputs/outputs?

**Comment by EmilSoleymani** [comment]
Many of these points were fixed in #126 , some were invalidated during our meeting (related to Julia OO limitations - see [Design.pdf](https://github.com/smiths/vdisp/blob/main/docs/Design/Design.pdf)), but some have not been dealt with yet.

I will be closing this issue and opening multiple issues for each unvisited problem listed above.

**Comment by EmilSoleymani** [comment]
Issues #128 #129 #130 #131 and #132 were created to address unvisited problems listed above.
